builtin/logical/pki/path_acme_order.go

```diff
@@ -884,3 +884,91 @@ func parseOrderIdentifiers(data map[string]interface{}) ([]*ACMEIdentifier, erro
 
 	return identifiers, nil
 }
+
+func (b *backend) acmeTidyOrder(ac *acmeContext, accountId string, orderPath string, sc *storageContext, certTidyBuffer time.Duration) (wasTidied bool, err error) {
+	// First we get the order; note that the orderPath includes the account
+	// It's only accessed at acme/orders/<order_id> with the account context
+	// It's saved at acme/<account_id>/orders/<orderId>
+	entry, err := ac.sc.Storage.Get(ac.sc.Context, orderPath)
+	if err != nil {
+		return false, fmt.Errorf("error loading order: %w", err)
+	}
+	if entry == nil {
+		return false, fmt.Errorf("order does not exist: %w", ErrMalformed)
+	}
+	var order acmeOrder
+	err = entry.DecodeJSON(&order)
+	if err != nil {
+		return false, fmt.Errorf("error decoding order: %w", err)
+	}
+
+	// Determine whether we should tidy this order
+	shouldTidy := false
+	// It is faster to check certificate information on the order entry rather than fetch the cert entry to parse:
+	if !order.CertificateExpiry.IsZero() {
+		// This implies that a certificate exists
+		// When a certificate exists, we want to expire and tidy the order when we tidy the certificate:
+		if time.Now().After(order.CertificateExpiry.Add(certTidyBuffer)) { // It's time to clean
+			shouldTidy = true
+		}
+	} else {
+		// This implies that no certificate exists
+		// In this case, we want to expire the order after it has expired (+ some safety buffer)
+		if time.Now().After(order.Expires) {
+			shouldTidy = true
+		}
+	}
+	if shouldTidy == false {
+		return shouldTidy, nil
+	}
+
+	// Tidy this Order
+	// That includes any certificate acme/<account_id>/orders/orderPath/cert
+	// That also includes any related authorizations: acme/<account_id>/authorizations/<auth_id>
+
+	// First Authorizations
+	for _, authorizationId := range order.AuthorizationIds {
+		authorizationEntry, err := sc.Storage.Get(ac.sc.Context, getAuthorizationPath(accountId, authorizationId))
+		if err != nil {
+			// If we continue here, we could fail to delete an authorization, but delete the order
+			// That would leave it dangling forever more, so we need to return
+			return false, err
+		}
+		if authorizationEntry == nil {
+			// If there was a storage error (above) in a previous run, it's possible that we've already tidied this
+			// In this case, don't fail, but do log that this shouldn't happen:
+			b.Logger().Warn("Encountered Missing Authorization Entry During Tidy: %v referred to Auth %v , but that authorization was not found", orderPath, authorizationId)
+		}
+
+		// Check Authorization Has Expired
+		var authorization ACMEAuthorization
+		err = authorizationEntry.DecodeJSON(&authorization)
+		if err != nil {
+			return false, err
+		}
+		expires, err := authorization.GetExpires()
+		if err != nil {
+			return false, err
+		}
+		if time.Now().Before(expires) {
```

Comment:
I think I'd still remove the authorization here. Since the order has been marked for tidy, and the authorization's validity period is separate from both the default order expiry and potentially the cert lifetime, simply removing all authorizations seems best.

I think you can simplify this loop to just the delete tbh.